src/Qir/Runtime/lib/Simulators/OpenSystemsSimulator.cpp

```diff
@@ -0,0 +1,292 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include <assert.h>
+#include <bitset>
+#include <complex>
+#include <iostream>
+#include <memory>
+#include <vector>
+
+#include "QirRuntimeApi_I.hpp"
+#include "QSharpSimApi_I.hpp"
+#include "SimFactory.hpp"
+
+using namespace std;
+typedef long long int i64;
+
+// Define the API with the open systems simulator.
+extern "C"
+{
+    // NB: We disable clang-tidy rules for case conventions here, as the names
+    //     reflect the Rust naming conventions used in the opensim crate.
+    size_t init(size_t initialcapacity); // NOLINT(readability-identifier-naming)
+    i64 destroy(size_t sim_id); // NOLINT(readability-identifier-naming)
+    void dump_to_console(size_t sim_id); // NOLINT(readability-identifier-naming)
+    i64 x(size_t sim_id, size_t idx); // NOLINT(readability-identifier-naming)
+    i64 y(size_t sim_id, size_t idx); // NOLINT(readability-identifier-naming)
+    i64 z(size_t sim_id, size_t idx); // NOLINT(readability-identifier-naming)
+    i64 h(size_t sim_id, size_t idx); // NOLINT(readability-identifier-naming)
+    i64 s(size_t sim_id, size_t idx); // NOLINT(readability-identifier-naming)
+    i64 s_adj(size_t sim_id, size_t idx); // NOLINT(readability-identifier-naming)
+    i64 t(size_t sim_id, size_t idx); // NOLINT(readability-identifier-naming)
+    i64 t_adj(size_t sim_id, size_t idx); // NOLINT(readability-identifier-naming)
+    i64 cnot(size_t sim_id, size_t idx_control, size_t idx_target); // NOLINT(readability-identifier-naming)
+    i64 m(size_t sim_id, size_t idx, size_t* result_out); // NOLINT(readability-identifier-naming)
+    const char* lasterr(); // NOLINT(readability-identifier-naming)
+    const char* get_noise_model(size_t sim_id); // NOLINT(readability-identifier-naming)
+    i64 set_noise_model(size_t sim_id, const char* new_model); // NOLINT(readability-identifier-naming)
+    const char* get_current_state(size_t sim_id); // NOLINT(readability-identifier-naming)
+}
```

Comment:
I feel that there can be multiple entities (files) that may need this `extern "C"` declarations, but not the rest of this file.
If that's the case or possible scenario then you might want to consider extracting this `extern "C"` block into a separate header.

Reply:
Agreed. As noted at https://github.com/microsoft/qsharp-runtime/blob/72977df6a9ddbc17b9dcecc3dfb7fd355f348a81/src/Simulation/qdk_sim_rs/docs/c-api.md, it may make sense to even remove this entirely and replace it by a header generated automatically by `cbindgen`.

Reply:
Added to https://github.com/microsoft/qsharp-runtime/issues/713; if it's alright with you, I'll defer to when the QIR work is resumed to address separating the C API and relying on `cbindgen`.